browser/brave_ads/brave_ads_host.cc

```diff
@@ -0,0 +1,162 @@
+
+}  // namespace
+
+BraveAdsHost::BraveAdsHost(content::WebContents* web_contents)
+    : WebContentsObserver(web_contents) {
+  DCHECK(web_contents);
+}
+
+BraveAdsHost::~BraveAdsHost() {
+  DCHECK(!callback_);
+}
+
+void BraveAdsHost::RequestAdsEnabled(RequestAdsEnabledCallback callback) {
+  DCHECK(callback);
+
+  Profile* profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  DCHECK(profile);
+
+  const AdsService* ads_service = AdsServiceFactory::GetForProfile(profile);
+  brave_rewards::RewardsService* rewards_service =
+      brave_rewards::RewardsServiceFactory::GetForProfile(profile);
+  if (!rewards_service || !ads_service || !ads_service->IsSupportedLocale()) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  if (ads_service->IsEnabled()) {
+    std::move(callback).Run(true);
+    return;
+  }
+
+  if (callback_) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  Browser* browser = chrome::FindBrowserWithWebContents(web_contents());
+  if (!browser) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  callback_ = std::move(callback);
+
+  rewards_service->StartProcess(base::DoNothing());
+  rewards_service_observation_.Observe(rewards_service);
+  action_ui_observation_.Observe(extensions::BraveActionAPI::Get(browser));
+  initiated_browser_ = browser;
+
+  if (!ShowRewardsPopup(browser, profile)) {
+    Reset();
+    std::move(callback_).Run(false);
+  }
+}
+
+void BraveAdsHost::WebContentsDestroyed() {
+  Reset();
+
+  if (callback_) {
+    std::move(callback_).Run(false);
+  }
+}
+
+void BraveAdsHost::OnRewardsEnabled() {
+  DCHECK(callback_);
+
+  Reset();
+
+  std::move(callback_).Run(true);
+}
+
+void BraveAdsHost::OnAdsEnabled(brave_rewards::RewardsService* rewards_service,
+                                bool ads_enabled) {
+  DCHECK(rewards_service);
+  DCHECK(callback_);
+
+  Reset();
+
+  std::move(callback_).Run(ads_enabled);
+}
+
+void BraveAdsHost::OnActionUIClosed(Browser* browser,
+                                    const std::string& extension_id) {
+  DCHECK(callback_);
+  if (extension_id != brave_rewards_extension_id) {
+    return;
+  }
+
+  const Browser* current_browser =
+      chrome::FindBrowserWithWebContents(web_contents());
+  if (browser != current_browser && browser != initiated_browser_) {
+    return;
+  }
+
+  Reset();
+
+  Profile* profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  DCHECK(profile);
+
+  const AdsService* ads_service = AdsServiceFactory::GetForProfile(profile);
+  bool ads_enabled = false;
+  if (ads_service) {
+    ads_enabled = ads_service->IsEnabled();
+  }
+  std::move(callback_).Run(ads_enabled);
+}
+
+bool BraveAdsHost::ShowRewardsPopup(Browser* browser, Profile* profile) {
+  auto* extension_service =
+      extensions::ExtensionSystem::Get(profile)->extension_service();
+  if (!extension_service) {
+    return false;
+  }
+
+  static_cast<extensions::BraveComponentLoader*>(
```

Comment:
this is not a synchronous action, you can't just open the extension right after adding it. You have to wait for the extension to be enabled and loaded

Reply:
also this is called from several different places and it seems like it should be a method on the rewards service instead of duplicating this code with potential mistakes like this one.

Reply:
As Ivan pointed, OnExtensionLoaded() is called synchronously from AddExtension(). So this should work fine.

Reply:
Oh, because this is a component extension it doesn't have to load from disk. cc @petemill

Reply:
Actually it is a bit more complex: `OnExtensionLoaded()` is called synchronously for all extensions, but it doesn't mean e.g. the background process has started or something. AFAIU `ExtensionAction` is also inserted synchronously so we can trigger the popup (that would launch a separate renderer for itself).